sbncode/SBNEventWeight/Calculators/Geant4/Geant4WeightCalc.cxx

```diff
@@ -0,0 +1,492 @@
+    p_final_proc="dummy";
+    p_init_momentum=-9999;
+    p_final_momentum=-9999;
+    p_energies_el.clear();
+    p_sliceInts_el.clear();
+    p_energies_inel.clear();
+    p_sliceInts_inel.clear();
+    p_nElasticScatters=-9999;
+
+
+    const simb::MCParticle& p = *mcparticles.at(i);
+    p_PDG = p.PdgCode();
+    int mcpID = p.TrackId();
+    std::string EndProcess  = p.EndProcess();
+
+    double mass = 0.;
+    if( TMath::Abs(p_PDG) == 211 ) mass = 139.57;
+    else if( p_PDG == 2212 ) mass = 938.28;
+
+    // We only want to record weights for one type of particle (defined by fPDG from the fcl file), so skip other particles
+    if (p_PDG == fPdg){
+      // Get GEANT trajectory points: weighting will depend on position and momentum at each trajectory point so calculate those
+      std::vector<double> trajpoint_X;
+      std::vector<double> trajpoint_Y;
+      std::vector<double> trajpoint_Z;
+      std::vector<double> trajpoint_PX;
+      std::vector<double> trajpoint_PY;
+      std::vector<double> trajpoint_PZ;
+      std::vector<int> elastic_indices;
+
+      //Get the list of processes from the true trajectory
+      const std::vector< std::pair< size_t, unsigned char > > & processes = p.Trajectory().TrajectoryProcesses();
+      std::map< size_t, std::string > process_map;
+      for( auto it = processes.begin(); it != processes.end(); ++it ){
+        process_map[ it->first ] = p.Trajectory().KeyToProcess( it->second );
+      }
+
+      for( size_t i = 0; i < p.NumberTrajectoryPoints(); ++i ){
+        double X = p.Position(i).X();
+        double Y = p.Position(i).Y();
+        double Z = p.Position(i).Z();
+        geo::Point_t testpoint1 { X, Y, Z };
+        const TGeoMaterial* testmaterial1 = fGeometryService->Material( testpoint1 );
+        //For now, just going to reweight the points within the LAr of the TPC
+        // TODO check if this is right
+        if ( !strcmp( testmaterial1->GetName(), "LAr" ) ){
+          trajpoint_X.push_back( X );
+          trajpoint_Y.push_back( Y );
+          trajpoint_Z.push_back( Z );
+
+          trajpoint_PX.push_back( p.Px(i) );
+          trajpoint_PY.push_back( p.Py(i) );
+          trajpoint_PZ.push_back( p.Pz(i) );
+
+          auto itProc = process_map.find(i);
+          if( itProc != process_map.end() && itProc->second == "hadElastic" ){
+            //Push back the index relative to the start of the reweightable steps
+            elastic_indices.push_back( trajpoint_X.size() - 1 );
+            // if (fDebug) std::cout << "Elastic index: " << trajpoint_X.size() - 1 << std::endl;
+          }
+
+        }
+
+      } // end loop over trajectory points
+
+      // Now find daughters of the MCP
+      std::vector<int> daughter_PDGs;
+      std::vector<int> daughter_IDs;
+      for( int i_mcp = 0; i_mcp < p.NumberDaughters(); i_mcp++ ){
+        int daughterID = p.Daughter(i_mcp);
+        for (auto test_mcp : mcparticles){
+          if (test_mcp->TrackId() == daughterID){
```

Comment:
I thought `daughterID` was the location in the array of the `.Daughter(i_mcp)` and not the trackID:
https://internal.dunescience.org/doxygen/MCParticle_8cxx_source.html#l00112